### PR TITLE
gh-113174: Sync with importlib_metadata 7.0

### PR DESCRIPTION
**Pull request by jaraco**
* Issue: gh-113174

### CODE REVIEW

**Comment by hugovk** [comment]
> Sync with importlib_metadata 7.0, including improved type annotations, fixed
issue with symlinked packages in ``package_distributions``, added
``EntryPoints.__repr__``, introduced the ``diagnose`` script, added
``Distribution.origin`` property, and removed deprecated ``EntryPoint``
access by numeric index (tuple behavior).

Will there be docs changes in this PR or another?

**Comment by jaraco** [comment]
> Will there be docs changes in this PR or another?

I've added docs in b610b6d. I chose not to mention some of the changes as they seemed incidental enough not to affect most users.

File: Doc/library/importlib.metadata.rst [hunk 1]
```diff
@@ -171,16 +171,18 @@ group.  Read `the setuptools docs
...
-
-The "selectable" entry points were introduced in ``importlib_metadata``
-3.6 and Python 3.10. Prior to those changes, ``entry_points`` accepted
-no parameters and always returned a dictionary of entry points, keyed
-by group. With ``importlib_metadata`` 5.0 and Python 3.12,
-``entry_points`` always returns an ``EntryPoints`` object. See
-`backports.entry_points_selectable <https://pypi.org/project/backports.entry-points-selectable>`_
-for compatibility options.
-
+.. versionchanged::3.12
```
Comment (hugovk) [inline Doc/library/importlib.metadata.rst]: Need spaces here for these to render:

```suggestion
.. versionchanged:: 3.12
```

https://cpython-previews--113175.org.readthedocs.build/en/113175/library/importlib.metadata.html#entry-points

File: Doc/library/importlib.metadata.rst [hunk 2]
```diff
@@ -171,16 +171,18 @@ group.  Read `the setuptools docs
...
+   The "selectable" entry points were introduced in ``importlib_metadata``
+   3.6 and Python 3.10. Prior to those changes, ``entry_points`` accepted
+   no parameters and always returned a dictionary of entry points, keyed
+   by group. With ``importlib_metadata`` 5.0 and Python 3.12,
+   ``entry_points`` always returns an ``EntryPoints`` object. See
+   `backports.entry_points_selectable <https://pypi.org/project/backports.entry-points-selectable>`_
+   for compatibility options.
+
+.. versionchanged::3.13
```
Comment (hugovk) [inline Doc/library/importlib.metadata.rst]:
```suggestion
.. versionchanged:: 3.13
```

File: Doc/library/importlib.metadata.rst [hunk 3]
```diff
@@ -342,9 +344,17 @@ instance::
...
+For editable packages, an origin property may present :pep:`610`
+metadata::
+
+    >>> dist.origin.url
+    'file:///path/to/wheel-0.32.3.editable-py3-none-any.whl'
+
 The full set of available metadata is not described here.
 See the `Core metadata specifications <https://packaging.python.org/en/latest/specifications/core-metadata/#core-metadata>`_ for additional details.
 
+.. versionadded::3.13
```
Comment (hugovk) [inline Doc/library/importlib.metadata.rst]:
```suggestion
.. versionadded:: 3.13
```

**Review by hugovk — COMMENTED** [review]
Would this also benefit from an entry in What's New in Python 3.13?

File: Doc/library/importlib.metadata.rst (hunk 1 above)
Comment (jaraco) [reply Doc/library/importlib.metadata.rst]: Thanks. I was inspecting the output and confused I couldn't find these sections. I checked the build logs for warnings and there were none. Thanks for catching the issue. I've rewritten the docs commit.

**Comment by jaraco** [comment]
> Would this also benefit from an entry in What's New in Python 3.13?

Perhaps. I'd consider mentioning the `EntryPoint.__getitem__` removal and new `Distribution.origin` property, though I expect most users not to care much. Overall, I'm thinking it'd be more noise than value, so I'll propose not to add anything, but I'd be happy to add something if anyone feels otherwise.

File: Doc/library/importlib.metadata.rst [hunk 4]
```diff
@@ -171,16 +171,18 @@ group.  Read `the setuptools docs
...
+   3.6 and Python 3.10. Prior to those changes, ``entry_points`` accepted
+   no parameters and always returned a dictionary of entry points, keyed
+   by group. With ``importlib_metadata`` 5.0 and Python 3.12,
+   ``entry_points`` always returns an ``EntryPoints`` object. See
+   `backports.entry_points_selectable <https://pypi.org/project/backports.entry-points-selectable>`_
+   for compatibility options.
+
+.. versionchanged:: 3.13
+   ``EntryPoint`` objects no longer present a tuple-like interface
+   (``__getitem__``).
```
Comment (AlexWaygood) [inline Doc/library/importlib.metadata.rst]: You could possibly link to the datamodel docs for the method here:

```suggestion
   (:meth:`~object.__getitem__`).
```

**Review by AlexWaygood — COMMENTED** [review]
I agree with @hugovk that the removal of `EntryPoint.__getitem__` should be mentioned in "What's New". It doesn't need to be a prominent mention near the top of the document, but we should try to make sure that all backwards-incompatible changes are mentioned somewhere in that document, in my opinion.

File: Doc/whatsnew/3.13.rst [hunk 5]
```diff
@@ -1001,6 +1001,10 @@ importlib
   for migration advice.
   (Contributed by Jason R. Coombs in :gh:`106532`.)
 
+* Remove deprecated ``__getattr__`` access for
```
Comment (AlexWaygood) [inline Doc/whatsnew/3.13.rst]:
```suggestion
* Remove deprecated :meth:`~object.__getitem__` access for
```

**Comment by jaraco** [comment]
Today I learned about the syntax ([docs](https://www.sphinx-doc.org/en/master/usage/domains/index.html#cross-referencing-syntax)). The exclamation point means to ignore EntryPoint because it isn't documented.

**Comment by jaraco** [comment]
Thanks for the review. Merging, but happy to revisit if needed.

**Comment by ambv** [comment]
There's some tests that are failing on Windows for this change. The build that introduced the failure is the one with this PR merged:
https://buildbot.python.org/all/#/builders/1079/builds/3009

**Comment by ambv** [comment]
I worked around the problem by enabling [SeCreateSymbolicLinkPrivilege](https://docs.python.org/3/library/os.html#os.symlink) for the buildbot user, but ultimately I believe the test should be skipped for users without this privilege set.

**Pull request merged**